multirpc/subpub/stream.go

```diff
@@ -10,45 +10,48 @@ import (
 )
 
 func (ps *SubPub) handleStream(stream network.Stream) {
+	// 'stream' will stay open until you close it (or the other side closes it).
+
+	log.Infof("connected to peer %s: %+v", stream.Conn().RemotePeer(), stream.Conn().RemoteMultiaddr())
+
+	// create a buffer to allow queueing up to 8 outgoing messages per peer.
+	// This allows us to concurrently spread messages without blocking,
+	// falling back to dropping messages if any peer is slow or disconnects.
 	peerClosed := make(chan bool)
+	writeBuffer := make(chan []byte, 8)
 
-	// First, ensure that any messages read from the stream are sent to the
-	// SubPub.Reader channel.
+	// Any messages read from the stream are sent to the SubPub.Reader channel.
 	go ps.readHandler(peerClosed, stream)
 
-	// Second, ensure that, from now on, any broadcast message is sent to
-	// this stream as well.
-	// Allow up to 8 queued broadcast messages per peer. This allows us to
-	// concurrently spread broadcast messages without blocking, falling back
-	// to dropping messages if any peer is slow or disconnects.
-	// TODO(mvdan): if another peer opens a stream with us, to just send
-	// us a single message (unicast), it's wasteful to also send broadcasts
-	// back via that stream.
+	go ps.writeHandler(peerClosed, writeBuffer, stream)
 
-	write := make(chan []byte, 8)
+	// This ps.Peers is probably not needed if we implement GossipSub as broadcast
```

Comment:
i had the feeling ps.Peers would become useless but later found out it's used for other things, among them unicasts.
i still have the feeling this whole ps.Peers could be dropped and reimplemented with GossipSub since GossipSub already tracks which "peers" are in a topic.
it would mean a significant refactoring of subpub, tho.